proto/kvrpcpb.proto

```diff
@@ -1534,3 +1534,39 @@ message TiFlashSystemTableRequest {
 message TiFlashSystemTableResponse {
     bytes data = 1 [(gogoproto.customtype) = "github.com/pingcap/kvproto/pkg/sharedbytes.SharedBytes", (gogoproto.nullable) = false];
 }
+
+// Flush is introduced from the pipelined DML protocol. 
+// A Flush request writes some keys and values to TiKV, storing in LOCK and DEFAULT CF, just like a Prewrite request.
+message FlushRequest {
+    Context context = 1;
+    repeated Mutation mutations = 2;
+    bytes primary_key = 3;
+    uint64 start_ts = 4;
+    uint64 min_commit_ts = 5;
+    // generation of the flush request. It is a monotonically increasing number in each transaction.
+    uint64 generation = 6;
+    uint64 lock_ttl = 7;
+    AssertionLevel assertion_level = 8;
+}
+
+message FlushResponse {
+    errorpb.Error region_error = 1;
+    repeated KeyError errors = 2;
+    ExecDetailsV2 exec_details_v2 = 3;
+}
+
+// BufferBatchGet is introduced from the pipelined DML protocol.
+// It is similar to a BatchGet request, except that it can only read the data that has been flushed by itself.
+message BufferBatchGetRequest {
+    Context context = 1;
+    repeated bytes keys = 2;
+    uint64 version = 3;
```

Comment:
Shall we add a variable to control whether enable snapshot read after buffer read miss? This may save one RPC.

Reply:
A similar question, if it's designed to support both buffer-read and snapshot read, or do we need another interface to implement "UnionStoreBatchGet" at tikv side?

Reply:
> Shall we add a variable to control whether enable snapshot read after buffer read miss? This may save one RPC.

We can do it later when we are about to implement this. We don't have to finalize all details in one PR.

>A similar question, if it's designed to support both buffer-read and snapshot read, or do we need another interface to implement "UnionStoreBatchGet" at tikv side?

`BufferBatchGet` only serves buffer read. Its semantics is "reading the content offloaded to TiKV in the membuffer"
Snapshot read does not change.
Something like `UnionStoreBatchGet` is not needed, as `unionstore.get` is simply implemented as "buffer.get, if not found, try snapshot.get". This implementation does not have to change at all. 
The only thing we need to change is how we do buffer read. Does that sound straightforward and reasonable?
```
Previous:
membuffer.get = memdb.get
snapshot.get = tikv.get
unionstore.get = membuffer.get +? snapshot.get = memdb.get +? tikv.get

Now:
membuffer.get = mutable_memdb.get +? immutable_memdb.get +? tikv.buffer_batch_get
snapshot.get = tikv.get
unionstore.get = membuffer.get +? snapshot.get = mutable_memdb.get +? immutable_memdb.get +? tikv.buffer_batch_get +? tikv.get
```